operators/src/source/gdal_source/mod.rs

```diff
@@ -144,6 +163,120 @@ pub struct GdalDatasetGeoTransform {
     pub y_pixel_size: f64,
 }
 
+impl GdalDatasetGeoTransform {
+    /// Produce the `SpatialPartition` anchored at the datasets origin with a size of x * y pixels. This method handles non-standard pixel sizes.
+    pub fn spatial_partition(&self, x_size: usize, y_size: usize) -> SpatialPartition2D {
+        // the opposite y value (y value of the non origin edge)
+        let opposite_coord_y = self.origin_coordinate.y + self.y_pixel_size * y_size as f64;
+
+        // if the y-axis is negative then the origin is on the upper side.
+        let (upper_y, lower_y) = if self.y_pixel_size.is_sign_negative() {
+            (self.origin_coordinate.y, opposite_coord_y)
+        } else {
+            (opposite_coord_y, self.origin_coordinate.y)
+        };
+
+        let opposite_coord_x = self.origin_coordinate.x + self.x_pixel_size * x_size as f64;
+
+        // if the y-axis is negative then the origin is on the upper side.
+        let (left_x, right_x) = if self.x_pixel_size.is_sign_positive() {
+            (self.origin_coordinate.x, opposite_coord_x)
+        } else {
+            (opposite_coord_x, self.origin_coordinate.x)
+        };
+
+        SpatialPartition2D::new_unchecked(
+            Coordinate2D::new(left_x, upper_y),
+            Coordinate2D::new(right_x, lower_y),
+        )
+    }
+
+    /// Transform a `Coordinate2D` into a `GridIdx2D`
+    #[inline]
+    pub fn coordinate_to_grid_idx_2d(&self, coord: Coordinate2D) -> GridIdx2D {
+        // TODO: use an epsilon here?
+        let grid_x_index =
+            ((coord.x - self.origin_coordinate.x) / self.x_pixel_size).floor() as isize;
+        let grid_y_index =
+            ((coord.y - self.origin_coordinate.y) / self.y_pixel_size).floor() as isize;
+
+        [grid_y_index, grid_x_index].into()
+    }
+
+    fn spatial_partition_to_read_window(
+        &self,
+        spatial_partition: &SpatialPartition2D,
+    ) -> GdalReadWindow {
+        // World coordinates and pixel sizes use float values. Since the float imprecision might cause overflowing into the next pixel we use an epsilon to correct values very close the pixel borders. This logic is the same as used in [`GeoTransform::grid_idx_to_pixel_upper_left_coordinate_2d`].
+        const EPSILON: f64 = 0.000_001;
+        let epsilon: Coordinate2D =
+            (self.x_pixel_size * EPSILON, self.y_pixel_size * EPSILON).into();
+
+        /*
+        The read window is relative to the transform of the gdal dataset. The `SpatialPartition` is oriented at axis of the spatial SRS. This usually causes this situation:
+
+        The gdal data is stored with negative pixel size. The "ur" coordinate of the `SpatialPartition` is neareest to the origin of the gdal raster data.
+        ur                      ul
```

Comment:
why is `ur` in the left?

Reply:
[44bd6b5](https://github.com/geo-engine/geoengine/pull/602/commits/44bd6b52f5b3a1e8f344edb0cd17692038662067)